cloudsmith/data_source_package_download.go

```diff
@@ -0,0 +1,150 @@
+package cloudsmith
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+// Hit the API with function PackagesListExecute with the following query: name and version (defaults to latest unless otherwise specified)
+func getPackageURL(pc *providerConfig, namespace string, repository string, query string, packageName string, packageVersion string) (string, string, error) {
+	req := pc.APIClient.PackagesApi.PackagesList(pc.Auth, namespace, repository)
+
+	queryString := query
+
+	if packageVersion != "" {
+		queryWithVersion := fmt.Sprintf("%s version:%s", query, packageVersion)
+		queryString = queryWithVersion
+	}
```

Comment:
why treat version specially if all we're doing is appending it to the query string anyway? would it be simpler to just allow the user to specify the full query they need? could be much more flexible that way

Reply:
it does feel unnecessary to me tbh, especially in the context of a TF provider, the query field alone leaves things fully flexible and even lets users play with searches/filters in the UI to build the query they want

i expect a big usecase for this might also be pulling packages by hash or other params rather than version.. i can think of a bunch of cases where version alone probably wouldn't be enough e.g. pulling a debian package you probably want to specify a distro, a python package you might want to specify wheel or sdist, etc

it's easy to add version later in a backwards compatible way if we decide that's desirable, but i'm not sure the reverse is true